src/libaktualizr/primary/sotauptaneclient.cc

```diff
@@ -1225,51 +1228,54 @@ void SotaUptaneClient::rotateSecondaryRoot(Uptane::RepositoryType repo, Uptane::
   }
 }
 
-// TODO(OTA-4342): the function can't currently return any errors. The problem of error reporting from
-// secondaries should be solved on a system (backend+frontend) error.
 // TODO: the function blocks until it updates all the secondaries. Consider non-blocking operation.
-void SotaUptaneClient::sendMetadataToEcus(const std::vector<Uptane::Target> &targets) {
+bool SotaUptaneClient::sendMetadataToEcus(const std::vector<Uptane::Target> &targets) {
   Uptane::RawMetaPack meta;
   if (!storage->loadLatestRoot(&meta.director_root, Uptane::RepositoryType::Director())) {
     LOG_ERROR << "No Director Root metadata to send";
-    return;
+    return false;
   }
   if (!storage->loadNonRoot(&meta.director_targets, Uptane::RepositoryType::Director(), Uptane::Role::Targets())) {
     LOG_ERROR << "No Director Targets metadata to send";
-    return;
+    return false;
   }
   if (!storage->loadLatestRoot(&meta.image_root, Uptane::RepositoryType::Image())) {
     LOG_ERROR << "No Image repo Root metadata to send";
-    return;
+    return false;
   }
   if (!storage->loadNonRoot(&meta.image_timestamp, Uptane::RepositoryType::Image(), Uptane::Role::Timestamp())) {
     LOG_ERROR << "No Image repo Timestamp metadata to send";
-    return;
+    return false;
   }
   if (!storage->loadNonRoot(&meta.image_snapshot, Uptane::RepositoryType::Image(), Uptane::Role::Snapshot())) {
     LOG_ERROR << "No Image repo Snapshot metadata to send";
-    return;
+    return false;
   }
   if (!storage->loadNonRoot(&meta.image_targets, Uptane::RepositoryType::Image(), Uptane::Role::Targets())) {
     LOG_ERROR << "No Image repo Targets metadata to send";
-    return;
+    return false;
   }
 
-  for (auto targets_it = targets.cbegin(); targets_it != targets.cend(); ++targets_it) {
-    for (auto ecus_it = targets_it->ecus().cbegin(); ecus_it != targets_it->ecus().cend(); ++ecus_it) {
-      const Uptane::EcuSerial ecu_serial = ecus_it->first;
-
+  bool put_meta_succeed = true;
+  for (const auto &target : targets) {
+    for (const auto &ecu : target.ecus()) {
+      const Uptane::EcuSerial ecu_serial = ecu.first;
       auto sec = secondaries.find(ecu_serial);
-      if (sec != secondaries.end()) {
-        /* Root rotation if necessary */
-        rotateSecondaryRoot(Uptane::RepositoryType::Director(), *(sec->second));
-        rotateSecondaryRoot(Uptane::RepositoryType::Image(), *(sec->second));
-        if (!sec->second->putMetadata(meta)) {
-          LOG_ERROR << "Sending metadata to " << sec->first << " failed";
-        }
+      if (sec == secondaries.end()) {
+        continue;
+      }
+
+      /* Root rotation if necessary */
+      rotateSecondaryRoot(Uptane::RepositoryType::Director(), *(sec->second));
+      rotateSecondaryRoot(Uptane::RepositoryType::Image(), *(sec->second));
+      if (!sec->second->putMetadata(meta)) {
+        LOG_ERROR << "Sending metadata to " << sec->first << " failed";
+        put_meta_succeed = false;
```

Comment:
Should we abort if any installation fails? Or is that a separate problem to solve since we don't handle atomicity well at present and that's a bigger topic?

Reply:
That's a good question. After further thinking, it might be worse than ignoring the check as it allows an hostile secondary to block a primary (and other secondaries) update, even if the metadata checks out from the PoV of the primary.

Reply:
Interesting point. I would think that would be okay, but we need to properly decide what to do in these cases and this is decidedly out of the scope of this work.

Reply:
My understanding is that even if we don't interrupt on the first putMetadata error anyway sendFirmware/install won't be called even for those Secondaries for which putMetdata was successful.
I agree with Patrick that it has to be discussed with a wider audience including product owners/managers, personally, I think aktualizr should carry on "metadata putting" and installation/update process if there is at least one "successful" case and then retry if there were non-verification/validation errors and then report a result to the backend...